crux/src/Crux.hs

```diff
@@ -233,83 +290,188 @@ data ProgramCompleteness
+          -- use the same float mode, so no mismatch here should be possible.
+          case testEquality floatRepr1 floatRepr2 of
+            Just Refl ->
+              doSimWithResults lang opts compRef glsRef pathSatSym execFeatures profInfo (proveGoalsOnline goalSym)
+            Nothing -> fail "Impossible: the argument interpretation produced two different float modes"
+
+    Left rsns -> fail ("Invalid solver configuration:\n" ++ unlines rsns)
+
+-- | Core invocation of the symbolic execution engine
+--
+-- This is separated out so that we don't have to duplicate the code sequence in
+-- 'runSimulator'.  The strategy used to ultimately discharge proof obligations
+-- is a parameter to allow this code to use either an online or offline solver
+-- connection.
+--
+-- The main work in this function is setting up appropriate solver frames and
+-- traversing the goals tree, as well as handling some reporting.
+doSimWithResults :: (Logs, IsSymInterface sym)
+                 => Language opts
+                 -> Options opts
+                 -> IORef ProgramCompleteness
+                 -> IORef (Seq.Seq (ProvedGoals (Either AssumptionReason SimError)))
+                 -> sym
+                 -> [GenericExecutionFeature sym]
+                 -> ProfData sym
+                 -> (forall ext . CruxOptions -> SimCtxt sym ext -> Maybe (Goals (LPred sym AssumptionReason) (LPred sym SimError)) -> IO (Maybe (Goals (LPred sym AssumptionReason) (LPred sym SimError, ProofResult (Either (LPred sym AssumptionReason) (LPred sym SimError))))))
+                 -- ^ The function to use to prove goals; this is intended to be
+                 -- one of 'proveGoalsOffline' or 'proveGoalsOnline'
+                 -> IO ( ProgramCompleteness
+                       , Seq.Seq (ProvedGoals (Either AssumptionReason SimError))
+                       )
+doSimWithResults lang opts@(cruxOpts, _) compRef glsRef sym execFeatures profInfo goalProver = do
+  frm <- pushAssumptionFrame sym
+  inFrame profInfo "<Crux>" $ do
+    simulate lang execFeatures opts sym emptyModel $ \(Result res) -> do
+      case res of
+        TimeoutResult {} -> do
+          sayWarn "Crux" "Simulation timed out! Program might not be fully verified!"
+          writeIORef compRef ProgramIncomplete
+        _ -> return ()
+      popUntilAssumptionFrame sym frm
+      let ctx' = execResultContext res
+      inFrame profInfo "<Prove Goals>" $ do
+        todo <- getProofObligations sym
+        proved <- goalProver cruxOpts ctx' todo
+        mgt <- provedGoalsTree ctx' proved
+        case mgt of
+          Nothing -> return ()
+          Just gt -> modifyIORef glsRef (Seq.|> gt)
+  when (simVerbose cruxOpts > 1) $ do
+    say "Crux" "Simulation complete."
+  when (isProfiling cruxOpts) $ writeProf profInfo
+  (,) <$> readIORef compRef <*> readIORef glsRef
+
+isProfiling :: CruxOptions -> Bool
+isProfiling cruxOpts = profileCrucibleFunctions cruxOpts || profileSolver cruxOpts
```

Comment:
All of the simulator refactoring above is great, and just makes things cleaner overall.